nox.py

```diff
@@ -94,6 +58,22 @@ def session_reqcheck(session):
         session.run('gcprepotools', command, reqfile)
 
 
+def collect_sample_dirs(start_dir, blacklist=set()):
+    """Recursively collects a list of dirs that contain tests."""
+    # Collect all the directories that have tests in them.
+    samples = []
+    for parent, subdirs, files in os.walk(start_dir):
+        if any(f for f in files if f[-8:] == '_test.py'):
+            # This dir has tests in it. Add it, and stop recursion
```

Comment:
py.test will recurse down the subdirs when it's invoked on a dir, so recursing down a dir we've already collected is redundant.

You might ask - why manually collect subdirs here if py.test does it already? I believe this is because we want to run each of our manually-collected subdirs in their own virtualenvs, which pytest doesn't do, but you can ask Jon for the definitive answer :-)